src/common/async.cpp

```diff
@@ -46,6 +49,13 @@ void Async::join() {
     }
 }
 
+Async::~Async() {
+    // WARNING: This MUST be here to make sure we break the loop before we
+    // stop the thread. Not doing that leads to undefined behavior.
+    break_loop();
+    join();
```

Comment:
> 1. Should you check if it's running before calling break_loop()? 

No. Break_loop() has no side effect if the loop is not running. (There was on purpose a regress test that checks that nothing bad happens if we're not running the loop and the thread.)

> 2. Why do you need to call join here, and which thread are you blocking?

I need to call `join()` to join the thread possibly started by `run_test()`. (Note that if the thread has not been started `join()` is a no-op.) More specifically, the reason by which I must perform both operations here is because such thread uses the `this` of the current object (you can probably imagine me mumbling something about using `Var<T>` head instead, which incidentally would be memory safe by construction). As such, `this` is shared and we must make sure we stop also the background thread when the `Async` object goes out of scope. (Note that the default destructor of the `std::thread` would have joined in any case, but given that in here there is potential of doing mistakes, I've decided to join the thread explicitly.)